### DISCUSSION TITLE
How to use this helper tool to listen to NSWorkspace Notifications

### DISCUSSION THREAD

**Discussion by metacodes** [trilemma-dev/SwiftAuthorizationSample#14] [Q&A] [answered]
Hello, firstly, thanks to your great work! Currently, I'm doing some POC for [this idea](https://github.com/lwouis/alt-tab-macos/issues/447#issuecomment-1112792851) base on your code. I've been able to run the program. But when I start to implement my ideas, such as listening to notification in helper, this program will report some errors, I refer to some documentation, but still no solution, I would like to ask you to see if there are any good ideas, I personally am an amateur in the development of Apple applications.
The following is my code:
https://github.com/metacodes/SwiftAuthorizationSample/blob/fdcc3b3c82e9cf49596bf48f06f73ff16c778826/SwiftAuthorizationHelperTool/main.swift#L30

I use `WorkspaceEvents.registerFrontAppChangeNote()` to register notification in the main function.
```Swift
import Foundation
import SecureXPC

NSLog("starting helper tool. PID \(getpid()). PPID \(getppid()).")
NSLog("version: \(try HelperToolInfoPropertyList.main().version.rawValue)")

// Command line arguments were provided, so process them
if CommandLine.arguments.count > 1 {
    // Remove the first argument, which represents the name (typically the full path) of this helper tool
    var arguments = CommandLine.arguments
    _ = arguments.removeFirst()
    NSLog("run with arguments: \(arguments)")
    
    if let firstArgument = arguments.first {
        if firstArgument == Uninstaller.commandLineArgument {
            try Uninstaller.uninstallFromCommandLine(withArguments: arguments)
        } else {
            NSLog("argument not recognized: \(firstArgument)")
        }
    }
} else if getppid() == 1 { // Otherwise if started by launchd, start up XPC server
    NSLog("parent is launchd, starting up XPC server")
    // register notification
    WorkspaceEvents.registerFrontAppChangeNote()

    let server = try XPCServer.forThisBlessedHelperTool()
    server.registerRoute(SharedConstants.allowedCommandRoute, handler: AllowedCommandRunner.run(message:))
    server.registerRoute(SharedConstants.uninstallRoute, handler: Uninstaller.uninstallFromXPC)
    server.registerRoute(SharedConstants.updateRoute, handler: Updater.updateHelperTool(atPath:))
    server.setErrorHandler { error in
        if case .connectionInvalid = error {
            // Ignore invalidated connections as this happens whenever the client disconnects which is not a problem
        } else {
            NSLog("error: \(error)")
        }
    }
    server.startAndBlock()
} else { // Otherwise started via command line without arguments, print out help info
    print("Usage: \(try CodeInfo.currentCodeLocation().lastPathComponent) <command>")
    print("\nCommands:")
    print("\t\(Uninstaller.commandLineArgument)\tUnloads and deletes from disk this helper tool and configuration.")
}
```
```Swift
class WorkspaceEvents {

    static func registerFrontAppChangeNote() {
        NSLog("starting registerFrontAppChangeNote 7")
        NSWorkspace.shared.notificationCenter.addObserver(self, selector: #selector(receiveFrontAppChangeNote(_:)), name: NSWorkspace.didActivateApplicationNotification, object: nil)
        NSWorkspace.shared.notificationCenter.addObserver(self, selector: #selector(receiveFrontAppLaunchNote(_:)), name: NSWorkspace.didLaunchApplicationNotification, object: nil)
        NSWorkspace.shared.notificationCenter.addObserver(self, selector: #selector(receiveFrontAppTerminateNote(_:)), name: NSWorkspace.didTerminateApplicationNotification, object: nil)
        NSLog("end registerFrontAppChangeNote 7")
    }

    // We add apps when we receive a didActivateApplicationNotification notification, not when we receive an apps launched, because any app will have an apps launched notification.
    // But we are only interested in apps that have windows. We think that since an app can be activated, it must have a window, and subscribing to its window event makes sense and is likely to work, even if it requires multiple retries to subscribe.
    // I'm not very sure if there is an edge case, but so far testing down the line has not revealed it.
    // When we receive the didActivateApplicationNotification notification, NSRunningApplication.isActive=true, even if the app is not the frontmost window anymore.
    // If we go to add the application when we receive the message of apps launched, at this time NSRunningApplication.isActive may be false, and try axUiElement.windows() may also throw an exception.
    // For those background applications, we don't receive notifications of didActivateApplicationNotification until they have their own window. For example, those menu bar applications.
    @objc static func receiveFrontAppChangeNote(_ notification: Notification) {
        NSLog("receiveFrontAppChangeNote")
        if let application = notification.userInfo?[NSWorkspace.applicationUserInfoKey] as? NSRunningApplication {
            NSLog("OS event \(notification.name.rawValue), \(application.bundleIdentifier), \(application.processIdentifier)")
        }
    }

    @objc static func receiveFrontAppLaunchNote(_ notification: Notification) {
        NSLog("receiveFrontAppLaunchNote")
        if let application = notification.userInfo?[NSWorkspace.applicationUserInfoKey] as? NSRunningApplication {
            NSLog("OS event \(notification.name.rawValue), \(application.bundleIdentifier), \(application.processIdentifier)")
        }
    }

    @objc static func receiveFrontAppTerminateNote(_ notification: Notification) {
        NSLog("receiveFrontAppTerminateNote")
        if let application = notification.userInfo?[NSWorkspace.applicationUserInfoKey] as? NSRunningApplication {
            NSLog("OS event \(notification.name.rawValue), \(application.bundleIdentifier), \(application.processIdentifier)")
        }
    }
}
```

But when the daemon launched, I got some errors.
```
Attempting to perform block on main runloop, but the main thread has exited. This message will only log once. Break on _CFRunLoopError_MainThreadHasExited to debug.
Attempting to wake up main runloop, but the main thread as exited. This message will only log once. Break on _CFRunLoopError_MainThreadHasExited to debug.
```
<img width="1639" alt="image" src="https://user-images.githubusercontent.com/3339872/168482802-20dd99c4-8ff2-40c9-bd33-cebb3ccb1f6c.png">
As I said above, the purpose of my POC is to listen to some application notifications in the background while creating AXRef. Any suggestions to solve this problem? Thanks!

**Comment by jakaplan** [comment, marked as answer]
`NSWorkspace` wouldn't be expected to work from a LaunchDaemon. `NSWorkspace` is part of AppKit which requires a GUI per-session bootstrap namespace. See [Technical Note TN2083's Layered Frameworks section](https://developer.apple.com/library/archive/technotes/tn2083/_index.html#//apple_ref/doc/uid/DTS10003794-CH1-SECTION23) for more details.

I'm not entirely sure what you're looking to do, but perhaps you want a LaunchAgent? LaunchDaemons don't interact with macOS's window server upon which all GUIs are based.

**Reply by metacodes** [reply]
Thanks for the answer! 

Currently what we want to do is we want to be able to start an XPCService to do some UI listening before the user logs in. This is because our XPCService needs to be launched before all UI applications are launched so that we can get all the AXUIElement of the application, but currently LaunchAgents cannot do that. LaunchAgents only starts when user logs into a graphical session, that's too late.

**Reply by jakaplan** [reply]
There are multiple misunderstandings and/or terminology issues here. What you're describing isn't possible because it's contradictory.

> Currently what we want to do is we want to be able to start an XPCService to do some UI listening before the user logs in.

[XPC services](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingXPCServices.html#//apple_ref/doc/uid/10000172i-SW6-SW1) can never do that. XPC services are sub-bundles of application bundles, only run (at most) during the lifetime of their containing application, and must be started via XPC by their containing application. They can only ever run as the user that's running the application (and by default are heavily sandboxed, but that can be disabled).

This sample is not for an XPC service, it's for a command line tool installed with `SMJobBless` that uses XPC as a communications mechanism. A service which uses XPC is not an XPC service. You can thank Apple for that rather bad naming choice.

>  so that we can get all the AXUIElement of the application

I know very little about macOS accessibility (nor is it related to this sample), so I can't be of any help to you in that regards.

>  LaunchAgents only starts when user logs into a graphical session, that's too late.

Prior to the user logging in, the session doesn't exist. Each user gets their own session. It can't be any sooner, as it does not exist before then.

It sounds like you have a larger design issue as `launchd` is intentional in *not* having ordering dependencies. It should be possible to query whatever state you need after your agent has been started.

**Reply by metacodes** [reply]
These explanations of yours are very useful to me, thank you very much!

I just found some documentation that might be helpful for this POC that I'm going to implement. I may be able to achieve the goal of the POC with PreLoginAgents.
Some references:
* [PreLoginAgents Demo](https://developer.apple.com/library/archive/samplecode/PreLoginAgents/Listings/PreLoginAgentCarbon__legacy__main_c.html)
* [Execute a script with GUI Access before login](https://apple.stackexchange.com/questions/372933/execute-a-script-with-gui-access-before-login)
